### ISSUE TITLE
Non-GW server cannot publish to other clusters

### ISSUE THREAD

**Issue by kapralius** [nats-io/nats-server#1146]
- [x] Feature Request or Change Proposal
 
Hi NATS team.

My first understanding of role of gateways in supercluster (after reading documuentation) was that it is a node providing connectivity to/from other clusters in the sense, that even non-GW servers from the same cluster can use it to gain extra-cluster connectivity.

After setting-up a testing supercluster topology I see that actually only the clients connected to GW-enabled nodes can publish to/subscribe from other clusters. 

Repro:
- Topology:   cluster UK ( non-gw_uk_1, gw_uk_1 ), cluster US (non-gw_us_1, gw_us_1)
- Client connections:
   - nats-sub -s non-gw_uk_1 data.>
   - nats-pub -s non-gw_us_1 data.xyz
- My expectation: nats_sub receives a message. 
- Actual behavior: nats_sub does not receive a message.

Is my setup correct and is this behavior expected? 

If so, I consider it as unfavorable from couple of reasons:

A) In multi-geolocation systems consisting of interconnected DCs, the nodes responsible for **interconnection to other DCs might be within DMZs** or other firewall protection, so only these nodes can connect to other DCs. From this reason, requiring each nats-server to connect to some other-cluster counterpart is problematic. 

B) In the topology of the system brokers (NATS servers) might be **dedicated to connect specific kinds of clients or to be subject of some scaling decisions**. Example: "broker concentrating frontend applications",  "broker dedicated for single client expected to generate/receive super-high-throughputs", "broker placed on low-performance hardware expected to dispatch low throughputs"... However, if any broker can potentially become a random "victim" being forced to perform a dispatching of messages of any other brokers from other clusters then any scaling decisions or dedication of brokers is useless.   

It must be noted, that I understand that this is a side effect of more general trait/design decision of NATS. Other effet of it is that in the case of failure, any client can potentially be redirected to other NATS broker within the cluster (not being redirected back even after the old broker is back online). This is also unfavorable in my case, but maybe it's making sense for others. 

C) I'm having trouble to imagine a scenario when this behavior of "non-GW node in supercluster unreachable to/from other clusters" is favorable. **It seems to me that if this is ever configured, it must be always a misconfiguration**, especially considering that in case of failure any client can be automatically redirected from GW-enable to non-GW server, so the client would transparently lose its former connectivity - is this right?

**Comment by derekcollison** [comment]
Every node in a cluster forms exactly one outbound GW connection to the remote clusters. They will also have [0-N] inbound connections from the remote cluster. A NATS client can connect to any server in any cluster and work properly.

https://nats-io.github.io/docs/gateways/
https://nats-io.github.io/docs/gateways/gateway.html

**Comment by kozlovic** [comment]
> that even non-GW servers from the same cluster can use it to gain extra-cluster connectivity

No that is not correct. Due to internal gossiping, a node that happens to not have the actual GW configuration should be notified by its peers in the cluster about that remote gateway and create an outbound connection there. This is usually used more in the case of auto-discovery of a gateway completely unknown to a cluster. For instance you have cluster A that has GW configuration for B. But there is a cluster C that now connects to B. B will inform C about A and C will auto connect to A (assuming that there is no configuration to actually reject unknown gateways).

In other words, if a node in a cluster is configured with a GW section, the other nodes in that same cluster should have the same configuration. But if they don't they should still be able to connect to the other cluster thanks to the gossip protocol. However, if the node being made aware of the URL of the remote cluster to connect to cannot issue a TCP connection, it will likely be reported in the log.

**Comment by kozlovic** [comment]
@kapralius Did we answer your questions? Should we close this issue or do you have more questions? Thanks!

**Comment by kapralius** [comment]
Thanks for your answers that clarified my misconception in how inter-cluster topology works.

What I did not comprehend from you answers until experimenting and checking source codes was how to avoid having all nodes acting as gateways. What I wanted was to setup the system with only part of the nodes being dedicated for receival of traffic from other clusters and dispatching it within the local cluster.

I found that a server without gateway section in the configuration (incl. name and listen endpoint defined) cannot ever connect and send any traffic to GWs in other clusters. So messages published from it will never be delivered to other clusters (see my minum example below to illustrate). Auto-discovery  will not help here, because such node actively refuses GW notification messages from other nodes on error "Received information about gateway %q from %s, but gateway is not configured".

If my understanding is true then sentence by @derekcollison:
>  Every node in a cluster forms exactly one outbound GW connection to the remote clusters.

Should really be:
> Every **GW** node in a cluster forms exactly one outbound GW connection to the remote clusters.

And this does not work as well:
> In other words, if a node in a cluster is configured with a GW section, the other nodes in that same cluster should have the same configuration. But if they don't they should still be able to connect to the other cluster thanks to the gossip protocol

In this light, also the following description from documentation seems to be misleading:
> Typically all cluster nodes will also be gateway nodes, but this is not a requirement.

... because, having a non-GW node within a supercluster is allowed configuration, but such server (and its clients) do not have connectivity to the part of the system. So for any practical purposes *all* nodes need to be gateway nodes.


So the conclusion for my desired setup was to make all the nodes having a gateway configuration section but only male part of the nodes to be "public gateways" - ie gateways that have URL listed in "gateways {}" configuration of all nodes in the supercluster.


This is a minimum setup demonstrating that messages published by non-GW node are not distributed to other cluster: 

	UK.non-GW.config:
		port: 1000
		cluster: {
				listen: localhost:5555
				routes: [ nats-route://localhost:5555 ]
		}

	UK.GW.config:
		port: 1001
		cluster: {
				listen: localhost:5556
				routes: [ nats-route://localhost:5555 ]
		}
		gateway {
			name: "UK-GW"
			listen: "localhost:7222"
			gateways: [
				{name: "UK-GW", url: "nats://localhost:7222"},
				{name: "US-GW", url: "nats://localhost:7223"},
			]
		}
		
	US.GW.config: 
		port: 2000
		cluster: {
				listen: localhost:5558
				routes: [ nats-route://localhost:5557 ]
		}
		gateway: {
			name: "US-GW"
			listen: "localhost:7223"
			gateways: [
				{name: "UK-GW", url: "nats://localhost:7222"},
				{name: "US-GW", url: "nats://localhost:7223"},
			]
		}
		
run:

		nats-server.exe -c "UK.non-GW.config"
		nats-server.exe -c "UK.GW.config"
		nats-server.exe -c "US.GW.config"
		nats-sub -s nats://localhost:2000 "data.>"
		nats-pub -s nats://localhost:1000 "data.123" helloworld

The published message will not be delivered to the subscriber.

**Comment by derekcollison** [comment]
Every node should participate in GW connections, we don't want any SPOF. /cc @kozlovic

**Comment by derekcollison** [comment]
We consider it mis-configured if only some nodes in a cluster form GW connections.

**Comment by kozlovic** [comment]
@kapralius You cannot have nodes in a cluster that do not have gateways enabled if other do. Well, you can, but then this node won't participate, so should be considered a misconfiguration.
What we have eluded to previously, is that a node may not have a specific gateway info and still be able to connect due to gossip. But if gateway is completely disabled (no gateway{} in config) then yes, the node is not going to be able to create outbound or accept inbounds and again, this should be considered misconfiguration.

**Comment by derekcollison** [comment]
I wonder if we should detect that @kozlovic and at least print warning etc?

**Comment by kozlovic** [comment]
We do (https://github.com/nats-io/nats-server/blob/master/server/route.go#L377), and user has mentioned that:

> Auto-discovery will not help here, because such node actively refuses GW notification messages from other nodes on error "Received information about gateway %q from %s, but gateway is not configured".

**Comment by longwallstreet** [comment]
We encountered the same problem with @kapralius @derekcollison 
In the actual business scenario, not all nodes in cluster UK can connect to cluster US node due to firewall restrictions. 
So in the current assumption, Non-GW node can only subscribe message from other cluster but the published message will not be delivered to other cluster.

**Comment by derekcollison** [comment]
We designed the GW system to not introduce SPOFs. In your case, why not just run one server then?

**Comment by kapralius** [comment]
@derekcollison - I'm not sure I'm getting the underlying reason why the abilities: 
- "I can use gateways of other clusters to publish my data" and 
- "I serve as gateway to recieve traffic from other clusters and dispatch it with my local cluster"

... were coupled together in NATS GW design. 

How such decoupling would lead to SPOF? If I have 10 nodes in my cluster and I decide that 7 of them will be relieved from the gateway duties (because they have network accessibility restrictions, they run within network segment with lower bandwidth or they serve for sensitive clients that saturate their resources anyway) and remaining 3 nodes remain in GW role for this cluster, I don't see where is the single point of failure is. Of course, I can still misconfigure it and make single node in GW role which would be wrong - but it's clearly my fault and as mentioned above, NATS allows to be misconfigured already.

**Comment by kozlovic** [comment]
As for the reason for current design: if you "elect" to have only some nodes in the cluster to be GW aware, it means that other nodes in the given cluster need to route published messages to one of the node that has GW abilities so that messages can be sent across. 
It would mean that a non GW node has to know the list of GW nodes, pick one and if the route to that node breaks, pick another, etc.. The GW node receiving a message from a route would have to be able to differentiate between a simple routed message due to local interest versus a message that is meant to be sent across (which means a different protocol message to make that distinction). It also means extra hops since right now each node has an outbound GW connection so message can be sent from any node. So I guess we could have done it this way but we chose to make it such as all nodes are participating.

> NATS allows to be misconfigured already

As you can see from the code, I report as an [ERR] but have a TODO that says should this be [FTL] (which would cause process to exit). We can revisit that.

**Comment by vazir** [comment]
@kozlovic @derekcollison would you please share a supercluster configuration samples? Superclustering concept logically means that there might be a number of isolated NATS clusters, interconnected by the gateways. But what I have read here, means that all NATS nodes in the clusters, must be able to contact any other nodes in the remote cluster (which might not be possible due the firewalling, or other restrinctions, like operating the clusters on the private IP range and only gatewaying nodes of it having the public IPs).

**Comment by kozlovic** [comment]
@vazir I understand your point and it is true that there are situations where one side can create outbound connections and the other can't, but unfortunately, gateways were designed in such a way that each server must have an outbound gateway to one of the server in the remote cluster.
So I can't share a configuration example where that is not the case, unfortunately.

**Comment by bruth** [comment]
Closing as the original question has been answered.
